ksqldb-engine/src/main/java/io/confluent/ksql/physical/scalablepush/PushPhysicalPlanBuilder.java

```diff
@@ -0,0 +1,169 @@
+import io.confluent.ksql.planner.plan.KsqlBareOutputNode;
+import io.confluent.ksql.planner.plan.OutputNode;
+import io.confluent.ksql.planner.plan.PlanNode;
+import io.confluent.ksql.planner.plan.PullFilterNode;
+import io.confluent.ksql.planner.plan.PullProjectNode;
+import io.confluent.ksql.query.QueryId;
+import io.confluent.ksql.util.KsqlException;
+import io.confluent.ksql.util.PersistentQueryMetadata;
+import io.vertx.core.Context;
+import java.util.Objects;
+
+/**
+ * Traverses the logical plan top-down and creates a physical plan for pull queries.
+ * The pull query must access a table that is materialized in a state store.
+ * The logical plan should consist of Project, Filter and DataSource nodes only.
+ */
+// CHECKSTYLE_RULES.OFF: ClassDataAbstractionCoupling
+public class PushPhysicalPlanBuilder {
+  // CHECKSTYLE_RULES.ON: ClassDataAbstractionCoupling
+
+  private final ProcessingLogContext processingLogContext;
+  private final PersistentQueryMetadata persistentQueryMetadata;
+  private final Stacker contextStacker;
+  private final QueryId queryId;
+
+  public PushPhysicalPlanBuilder(
+      final ProcessingLogContext processingLogContext,
+      final PersistentQueryMetadata persistentQueryMetadata
+  ) {
+    this.processingLogContext = Objects.requireNonNull(
+        processingLogContext, "processingLogContext");
+    this.persistentQueryMetadata = Objects.requireNonNull(
+        persistentQueryMetadata, "persistentQueryMetadata");
+    this.contextStacker = new Stacker();
+    queryId = uniqueQueryId();
+  }
+
+  /**
+   * Visits the logical plan top-down to build the physical plan.
+   * @param logicalPlanNode the logical plan root node
+   * @return the root node of the tree of physical operators
+   */
+  public PushPhysicalPlan buildPushPhysicalPlan(
+      final LogicalPlanNode logicalPlanNode,
+      final Context context
+  ) {
+    PushDataSourceOperator dataSourceOperator = null;
+
+    final OutputNode outputNode = logicalPlanNode.getNode()
+        .orElseThrow(() -> new IllegalArgumentException("Need an output node to build a plan"));
+
+    if (!(outputNode instanceof KsqlBareOutputNode)) {
+      throw new KsqlException("Push queries expect the root of the logical plan to be a "
+          + "KsqlBareOutputNode.");
+    }
+    // We skip KsqlBareOutputNode in the translation since it only applies the LIMIT
+    PlanNode currentLogicalNode = outputNode.getSource();
+    AbstractPhysicalOperator prevPhysicalOp = null;
+    AbstractPhysicalOperator rootPhysicalOp = null;
+    while (true) {
+      AbstractPhysicalOperator currentPhysicalOp = null;
+      if (currentLogicalNode instanceof PullProjectNode) {
+        currentPhysicalOp = translateProjectNode((PullProjectNode)currentLogicalNode);
+      } else if (currentLogicalNode instanceof PullFilterNode) {
+        currentPhysicalOp = translateFilterNode((PullFilterNode) currentLogicalNode);
+      } else if (currentLogicalNode instanceof DataSourceNode) {
```

Comment:
Should we start naming `PullProjectNode` and `PullFilterNode` to `TransientProjectNode` and `TransientFilterNode` in the codebase now that they are being used in different types of queries? This would make the naming more generic

Reply:
Called it QueryProjectNode, per offline discussion.